File: pi-repomap/ts-src/modes/pairs.ts
export function renderPairs(pairs: Record<string, string[]>, _tokenBudget: number): string {
  if (Object.keys(pairs).length === 0) {
    return "# Test/source pairs\nNo likely pairs found";
  }

  const lines = ["# Test/source pairs"];
  for (const [source, tests] of Object.entries(pairs)) {
    lines.push(`- ${source}`);
    for (const test of tests.slice(0, 8)) {
      lines.push(`  - ${test}`);
    }
  }
  return lines.join("\n");
}

File: pi-repomap/src/scope/modes/pairs.py
from __future__ import annotations

from typing import Dict, List


def render(pairs: Dict[str, List[str]], token_budget: int) -> str:
    """Test/source pair mapping — which source file likely has which test file.

    Tedious to do in bash (name matching + directory traversal), one call here.
    """
    _ = token_budget

    if not pairs:
        return "# Test/source pairs\nNo likely pairs found"

    lines = ["# Test/source pairs"]
    for source, tests in pairs.items():
        lines.append(f"- {source}")
        for test in tests[:8]:
            lines.append(f"  - {test}")

    return "\n".join(lines)
